src/app/clusters/scenes-server/SceneHandlerImpl.cpp

```diff
@@ -16,10 +16,156 @@
+/// @param EmberAfDefaultAttributeValue & defaultValue
+/// @param len Length of the byte array
+/// @return uint64_t or int64_t Value
+/// @note The attribute table supports 8 bytes only for unsigned integers, 4 bytes is the maximum for signed integers
+template <typename Type>
+Type ConvertByteArrayToType(const EmberAfDefaultAttributeValue & defaultValue, uint16_t len)
+{
+    if (len <= 2)
+    {
+        return static_cast<Type>(defaultValue.defaultValue);
+    }
+    else
+    {
+        Type sValue = 0;
+        memcpy(&sValue, defaultValue.ptrToDefaultValue, sizeof(Type));
+        return app::NumericAttributeTraits<Type>::StorageToWorking(sValue);
+    }
+}
+
+/// CapAttributeID
+/// Cap the attribute value based on the attribute's min and max if they are defined,
+/// or based on the attribute's size if they are not.
+/// @param[in] aVPair   AttributeValuePairType
+/// @param[in] metadata  EmberAfAttributeMetadata
+///
+template <typename Type>
+void CapAttributeID(AttributeValuePairType & aVPair, const EmberAfAttributeMetadata * metadata)
+{
+    // Calculate the maximum value that can be represented with the given number of bytes
+    Type maxValue = 0;
+
+    // Check if the attribute type is signed
+    if (metadata->IsSignedIntegerAttribute())
+    {
+        maxValue = static_cast<Type>((1ULL << (emberAfAttributeSize(metadata) * 8 - 1)) - 1);
+    }
+    else
+    {
+        maxValue = static_cast<Type>((1ULL << (emberAfAttributeSize(metadata) * 8)) - 1);
+    }
+
+    // Check metadata for min and max values
+    if (metadata->HasMinMax())
+    {
+        const EmberAfAttributeMinMaxValue * minMaxValue = metadata->defaultValue.ptrToMinMaxValue;
+        Type minVal = ConvertByteArrayToType<Type>(minMaxValue->minValue, emberAfAttributeSize(metadata));
+        Type maxVal = ConvertByteArrayToType<Type>(minMaxValue->maxValue, emberAfAttributeSize(metadata));
+
+        // Cap based on minimum value
+        if (minVal > static_cast<Type>(aVPair.attributeValue))
+        {
+            uint64_t sValue = 0;
+            memcpy(&sValue, &minVal, sizeof(Type));
+            aVPair.attributeValue = app::NumericAttributeTraits<uint64_t>::StorageToWorking(sValue);
+            // We assume the max is >= min therefore we can return
+            return;
+        }
+
+        // Adjust maxValue if greater than the meta data's max value
+        if (maxVal < maxValue)
+        {
+            maxValue = maxVal;
+        }
+    }
+
+    // Cap based on maximum value
+    if (metadata->IsSignedIntegerAttribute())
+    {
+        if (static_cast<int64_t>(aVPair.attributeValue) > static_cast<int64_t>(maxValue))
+        {
+            uint64_t sValue = 0;
+            memcpy(&sValue, &maxValue, sizeof(Type));
+            aVPair.attributeValue = app::NumericAttributeTraits<uint64_t>::StorageToWorking(sValue);
+        }
+    }
+    else
+    {
+        if (static_cast<uint64_t>(aVPair.attributeValue) > static_cast<uint64_t>(maxValue))
```

Comment:
Isn't `aVPair.attributeValue` already a `uint64_t`?  And why do we need to cast maxValue to anything?

Reply:
The casting is to ensure that they are both compared as uint64_t, I could have swore the type promotion didn't work as expected in a previous test but it seems I was wrong. We do need to cast maxValue for compare to silence warnings on different type comparison though.